pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/BaseMultipleSegmentsConversionExecutor.java

```diff
@@ -195,30 +200,9 @@ public List<SegmentConversionResult> executeTask(PinotTaskConfig pinotTaskConfig
     File tempDataDir = new File(new File(MINION_CONTEXT.getDataDir(), taskType), "tmp-" + UUID.randomUUID());
     Preconditions.checkState(tempDataDir.mkdirs());
     try {
-      List<File> inputSegmentDirs = new ArrayList<>();
-      int numRecords = 0;
-
-      for (int i = 0; i < downloadURLs.length; i++) {
-        String segmentName = segmentNames[i];
-        // Download and decompress the segment file
-        _eventObserver.notifyProgress(_pinotTaskConfig, "Downloading and decompressing segment from: " + downloadURLs[i]
-            + " (" + (i + 1) + " out of " + downloadURLs.length + ")");
-        File indexDir;
-        try {
-          indexDir = downloadSegmentToLocalAndUntar(tableNameWithType, segmentName, downloadURLs[i], taskType,
-              tempDataDir, "_" + i);
-        } catch (Exception e) {
-          LOGGER.error("Failed to download segment from download url: {}", downloadURLs[i], e);
-          _minionMetrics.addMeteredTableValue(tableNameWithType, MinionMeter.SEGMENT_DOWNLOAD_FAIL_COUNT, 1L);
-          _eventObserver.notifyTaskError(_pinotTaskConfig, e);
-          throw e;
-        }
-        inputSegmentDirs.add(indexDir);
-
-        reportSegmentDownloadMetrics(indexDir, tableNameWithType, taskType);
-        SegmentMetadataImpl segmentMetadata = new SegmentMetadataImpl(indexDir);
-        numRecords += segmentMetadata.getTotalDocs();
-      }
+      List<File> inputSegmentDirs = Collections.synchronizedList(new ArrayList<>());
```

Comment:
I think there are ordering requirements on inputSegmentDirs. These should be ordered based on order of correposponding downloadURLs / segmentNames. 

the controller orders the input segmentNames based on sequence id. 

And then upsert executor creates recordReaderFileConfigs in the same order and uses that in SegmentProcessorFramework. 

The order can be important for various edge cases. I am not fully sure if its ok to break the order but it would be safe to not break the order.

So if we can ensure the inputSegmentDirs are in the same order as downloadURLs / segmentNames that will help avoid regression.